File: C++/代码随想录/Linked_List/203. Remove Linked List Elements.cpp
// 203. Remove Linked List Elements
// https://leetcode.com/problems/remove-linked-list-elements/

#include "LinkedListUtils.hpp"

ListNode *removeElements(ListNode *head, int val) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode *prev = &dummy;
    ListNode *curr = head;

    while (curr != nullptr) {
        if (curr->val == val) {
            prev->next = curr->next;
            delete curr;
            curr = prev->next;
        } else {
            prev = curr;
            curr = curr->next;
        }
    }

    return dummy.next;
}

int main() {
    std::vector<int> values = {1, 2, 6, 3, 4, 5, 6};
    ListNode *head = createLinkedList(values); 
    ListNode *result = removeElements(head, 6);
    printList(result);

    return 0;
}

File: C++/代码随想录/Linked_List/203. Remove Linked List Elements.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        current = dummy

        while current.next:
            if current.next.val == val:
                current.next = current.next.next
            else:
                current = current.next

        return dummy.next

# 测试用例
def printList(head):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")

sol = Solution()

# 创建链表 1 -> 2 -> 6 -> 3 -> 4 -> 5 -> 6
head = ListNode(1, ListNode(2, ListNode(6, ListNode(3, ListNode(4, ListNode(5, ListNode(6)))))))

# 移除值为 6 的节点
printList(sol.removeElements(head, 6))  # 应该输出 1 -> 2 -> 3 -> 4 -> 5 -> None
